docs/src/main/asciidoc/extension-maturity-matrix.adoc

```diff
@@ -0,0 +1,169 @@
+
+=== Static initialization
+
+Do as much initialization as much as possible statically.
+This avoid runtime overhead.
+
+=== Replace reflection with generated bytecode
+
+Many Java libraries make heavy use of reflection to delay decisions to run-time. Quarkus aims to improve performance by moving logic to build time, reducing unnecessary dynamism.
+Extensions should aim to replace reflection with build-time code.
+This is enabled by
+   xref:writing-extensions.adoc#scanning-deployments-using-jandex[Jandex], an "offline reflection" library. It may also be necessary to do some bytecode transformation of existing libraries.
+
+For a case study of how to eliminate reflection and what the performance benefits turned out to be, see https://quarkus.io/blog/quarkus-metaprogramming/[reflectionless Jackson serialization]
+
+=== Virtual thread support
+
+Not every library is suitable for using with virtual threads, out of the box.
+xref:virtual-threads#why-not["Why not virtual threads everywhere?"] explains why.
+
+To get your library working properly with virtual threads, you should make sure the library is not pinning the carrier thread.
+ Quarkus has xref:virtual-threads.adoc#testing-virtual-thread-applications[test helpers to do these checks in an automated way].
+ For dispatching work, you should use the xref:virtual-threads.adoc#inject-the-virtual-thread-executor[virtual executor managed by Quarkus]. The link:https://quarkus.io/extensions/io.quarkus/quarkus-websockets-next/[WebSockets-next extension] uses the virtual dispatcher and smart dispatch, and is a good example to follow.
+
+=== Hot path performance optimization
+
+Although Quarkus offers some unique opportunities for extension performance, extension developers shouldn't forget https://www.linkedin.com/pulse/how-optimize-software-performance-efficiency-subcodevs/[the basics of performance optimization].
+
+=== Non-blocking internals
+
+Quarkus's reactive core is a key contributor to its excellent throughput and scalability. Extensions should consider adopting this model for their own internal operations.
+
+=== Add Mutiny-based APIs
+
+For maximum scalability, go beyond the reactive core and enable fully reactive programming, using Mutiny. Most projects that support a reactive programming model offer two distinct extensions, a `-reactive` and a plain one.
+See, for example, https://quarkus.io/extensions/io.quarkiverse.quarkus-elasticsearch/quarkus-elasticsearch/[ElasticSearch] and https://quarkus.io/extensions/io.quarkiverse.quarkus-elasticsearch-reactive/quarkus-elasticsearch-reactive/[ElasticSearch Reactive] extensions.
+
+== Operations
+
+Developer joy is important, but so are observability, maintainability, and other operational considerations.
+Many of these characteristics come by default with the Quarkus framework or https://quarkus.io/extensions/io.quarkus/quarkus-opentelemetry/[observability-focussed extensions]. But extensions can do more.
```

Comment:
Extensions need to contribute _binders_ to expose their metrics.

(this may change in the future).

For spans - I don't know if we need to mention the context propagation (duplicated context propagation somewhere), typically, it's something missing in the AWS extension..

Reply:
I've added a whole section about tracing, since that seems important and something extensions might forget about. Well, and since we've got room, because the operations section is thin. 

I left binders out since that seems specific to Micrometer and @brunobat suggests we just not give any advice until the technology direction is settled. :)